pkg/util/winutil/servicemain/servicemain.go

```diff
@@ -0,0 +1,197 @@
+	// enters the SERVICE_STOPPED state.
+	// golang implements its own ServiceMain function which calls RegisterServiceCtrlHandlerEx, it
+	// is up to our Execute() function to handle ChangeRequest's and update SCM with the service status.
+	// https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-startservicectrldispatcherw
+	// https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-registerservicectrlhandlerexw
+	// https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nc-winsvc-lpservice_main_functiona
+	// https://learn.microsoft.com/en-us/windows/win32/services/writing-a-service-program-s-main-function
+	// https://learn.microsoft.com/en-us/windows/win32/services/writing-a-servicemain-function
+	err := svc.Run(s.service.Name(), &s)
+	if err != nil {
+		s.eventlog(messagestrings.MSG_SERVICE_FAILED, err.Error())
+		return
+	}
+
+	// svc.Run() can return before Execute() ends if there is an error, but since we trigger
+	// process exit when svc.Run() returns its okay if we leak the goroutine.
+	s.eventlog(messagestrings.MSG_SERVICE_STOPPED, s.service.Name())
+}
+
+// runTimeExitGate is used to ensure the service exits without error on short-lived successful stops
+//
+// On Windows, the Service Control Manager (SCM) requires that dependent services stop. This means that when running
+// `Restart-Service datadogagent`, Windows will try to stop the Process Agent, and then to be helpful it will immediately start it again.
+// However, if Process Agent is not configured to be running it will exit immediately, which `Restart-Service` will report as an error.
+// To avoid the error on a successful exit we must ensure that we are in the RUNNING state long enough for `Restart-Service` or other
+// tools to consider the restart successful.
+func runTimeExitGate() <-chan time.Time {
+	return time.After(5 * time.Second)
+}
+
+func (s *controlHandler) eventlog(msgnum uint32, arg string) {
+	winutil.LogEventViewer(s.service.Name(), msgnum, arg)
+}
+
+// Execute is called by golang svc.Run and is responsible for handling the control requests and state transitions for the service
+// golang.org/x/sys/windows/svc contains the actual control handler callback and status handle, and communicates with
+// our Execute() function via the provided channels.
+// https://learn.microsoft.com/en-us/windows/win32/services/service-status-transitions
+func (s *controlHandler) Execute(args []string, r <-chan svc.ChangeRequest, changes chan<- svc.Status) (ssec bool, errno uint32) {
+
+	// first thing we must do is inform SCM that we are SERVICE_START_PENDING.
+	changes <- svc.Status{State: svc.StartPending}
+
+	const runningCmdsAccepted = svc.AcceptStop | svc.AcceptShutdown | svc.AcceptPreShutdown
+
+	err := s.service.Init()
+	if err != nil {
+		if errors.Is(err, ErrCleanStopAfterInit) {
+			changes <- svc.Status{State: svc.Running, Accepts: runningCmdsAccepted}
+			<-runTimeExitGate()
+			return
+		}
```

Comment:
What is the point of that error? I mean from SCM reporting perspective. Would not it restarted the service because it exits without SCM service stopping? I saw few comments in other places but it is still not clear to me.

Previous code reported changes <- svc.Status{State: svc.Stopped} which seems to be intuitively at least a "right" thing to do.

Reply:
I think a bit summary at this code-line would be sufficient. I think I did understand your previous point in the code above but this code was out of the way and somewhat disjointed that connection did not occur to me.